### PR TITLE
[Data] Infer data type of partition values

### PR DESCRIPTION
**Pull request by bveeramani**
## Why are these changes needed?

Ray Data can extract partition information from file paths. For example, if a file path is `year=2000/file`, Ray Data can add a `"year"` column to the resulting dataset. 

Currently, all extracted values are represented as strings. For consistency with other libraries like PyArrow, this PR makes Ray Data infer the data type.  

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/data/datasource/partitioning.py [hunk 1]
```diff
@@ -165,6 +175,7 @@ def of(
         style: PartitionStyle = PartitionStyle.HIVE,
         base_dir: Optional[str] = None,
         field_names: Optional[List[str]] = None,
+        field_types: Optional[Dict[str, Union[int, float, bool, str]]] = None,
```
Comment (raulchen) [inline python/ray/data/datasource/partitioning.py:178]:
```suggestion
        field_types: Optional[Dict[str, PartitionDataType]] = None,
```

**Review by raulchen — APPROVED** [review]
What would happen if the input file paths have inconsistent inference results? Can you add a test to cover this?

File: python/ray/data/datasource/partitioning.py [hunk 2]
```diff
@@ -226,14 +240,24 @@ def __call__(self, path: str) -> Dict[str, str]:
...
+        partitions: Dict[str, str] = self._parser_fn(dir_path)
+
+        field_types = dict(self._scheme.field_types)
+        for key, value in partitions.items():
+            field_types.setdefault(key, _infer_data_type(value))
+
+        return {
+            key: _cast_value(value, field_types[key])
+            for key, value in partitions.items()
+        }
```
Comment (alexeykudinkin) [inline python/ray/data/datasource/partitioning.py:260]: This allows different paths to potentially deduce different partition schemas -- we need to enforce that schema is consistent

**Comment by bveeramani** [comment]
Discussed internally -- decided we won't infer types, and will default to string if not provided by user

**Pull request closed without merging**